impeller/base/allocation_size.h

```diff
@@ -0,0 +1,186 @@
+  ///
+  constexpr AllocationSize(uint64_t byte_size, FromBytesTag)
+      : bytes_(byte_size) {}
+
+  //----------------------------------------------------------------------------
+  /// @return     The byte size.
+  ///
+  constexpr uint64_t GetByteSize() const { return bytes_; }
+
+  //----------------------------------------------------------------------------
+  /// @return     The number of `Periods` of bytes.
+  ///
+  constexpr double GetSize() const {
+    return GetByteSize() / static_cast<double>(Period);
+  }
+
+  //----------------------------------------------------------------------------
+  /// @brief      Convert the allocation size from one unit to another.
+  ///
+  ///             Conversions are non-truncating.
+  ///
+  /// @tparam     AllocationSize  The allocation size to convert to.
+  ///
+  /// @return     The new allocation size.
+  ///
+  template <class AllocationSize>
+  constexpr AllocationSize ConvertTo() {
+    return AllocationSize{GetByteSize(), FromBytesTag::kFromBytes};
+  }
+
+  // The following relational operators can be replaced with a defaulted
+  // spaceship operator post C++20.
+
+  constexpr bool operator<(const AllocationSize& other) const {
+    return bytes_ < other.bytes_;
+  }
+
+  constexpr bool operator>(const AllocationSize& other) const {
+    return bytes_ > other.bytes_;
+  }
+
+  constexpr bool operator>=(const AllocationSize& other) const {
+    return bytes_ >= other.bytes_;
+  }
+
+  constexpr bool operator<=(const AllocationSize& other) const {
+    return bytes_ <= other.bytes_;
+  }
+
+  constexpr bool operator==(const AllocationSize& other) const {
+    return bytes_ == other.bytes_;
+  }
+
+  constexpr bool operator!=(const AllocationSize& other) const {
+    return bytes_ != other.bytes_;
+  }
+
+  // Explicit casts.
+
+  explicit constexpr operator bool() const { return bytes_ != 0u; }
+
+  // Arithmetic operators (overflows are caller error).
+
+  constexpr AllocationSize operator+(const AllocationSize& other) const {
+    return AllocationSize(bytes_ + other.GetByteSize(),
+                          FromBytesTag::kFromBytes);
+  }
+
+  constexpr AllocationSize operator-(const AllocationSize& other) const {
+    return AllocationSize(bytes_ - other.GetByteSize(),
+                          FromBytesTag::kFromBytes);
+  }
+
+  constexpr AllocationSize& operator+=(const AllocationSize& other) {
+    bytes_ += other.GetByteSize();
+    return *this;
+  }
+
+  constexpr AllocationSize& operator-=(const AllocationSize& other) {
+    bytes_ -= other.GetByteSize();
+    return *this;
+  }
+
+ private:
+  uint64_t bytes_ = {};
+};
+
+using Bytes = AllocationSize<1u>;
+
+using KiloBytes = AllocationSize<1'000u>;
+using MegaBytes = AllocationSize<1'000u * 1'000u>;
+using GigaBytes = AllocationSize<1'000u * 1'000u * 1'000u>;
+
+using KibiBytes = AllocationSize<1'024u>;
+using MebiBytes = AllocationSize<1'024u * 1'024u>;
+using GigiBytes = AllocationSize<1'024u * 1'024u * 1'024u>;
```

Comment:
It's not `gigibytes`, but `gibibytes` :) You're probably going to need to change it across the PR, since the typo came up in a few places.

Reply:
derp. Good catch. Fixed everywhere.